### PR TITLE
JSpecify subtyping checks for arrays

### PR DESCRIPTION
**Pull request by msridhar**
Add checks for proper array subtyping (including nullability of array contents) at pseudo-assignments and for method overriding.  We do not yet handle multi-dimensional arrays.  Note that JSpecify allows for covariant array subtyping with respect to nullability (see https://github.com/jspecify/jspecify/issues/65) which complicates the implementation a bit.  Also some hacks are required since javac does not propagate the array contents nullability annotation into the types of the relevant trees in all cases.

### CODE REVIEW

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java [hunk 1]
```diff
@@ -213,6 +213,146 @@ public void nullableAssignmentParameterArray() {
...
+            "class Test {",
+            "  static void test(@Nullable Integer[] nullableIntArr, Integer[] nonnullIntArr) {",
+            "    // legal",
+            "    Integer[] x1 = nonnullIntArr;",
+            "    // legal",
+            "    @Nullable Integer[] x2 = nullableIntArr;",
+            "    // legal (covariant array subtypes)",
+            "    x2 = nonnullIntArr;",
+            "    // BUG: Diagnostic contains: Cannot assign from type @Nullable Integer[] to type Integer[]",
+            "    x1 = nullableIntArr;",
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java:232]: To be clear, we don't try to infer nullability annotations for elements of local variables of array type or nullability of generics for object types, right? Or is it just that once assigned a `@NonNull Integer[]` then `x1` must be assume to be incompatible with `@Nullable Integer[]`?

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java [hunk 2]
```diff
@@ -213,6 +213,146 @@ public void nullableAssignmentParameterArray() {
...
+            "  }",
+            "  static void testPositiveTernaryMethodArgument(boolean t) {",
+            "    // BUG: Diagnostic contains: Conditional expression must have type",
+            "    Integer[] a = testPositive(t ? new Integer[0] : new @Nullable Integer[0], t);",
+            "  }",
+            "  static @Nullable Integer[] testNegative(boolean n) {",
+            "    @Nullable Integer[] t1 = n ? new @Nullable Integer[0] : new @Nullable Integer[0];",
+            "    @Nullable Integer[] t2 = n ? new Integer[0] : new @Nullable Integer[0];",
+            "    return n ? new @Nullable Integer[0] : new @Nullable Integer[0];",
+            "  }",
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java:351]: test negative arg case missing (mostly for the sake of symmetry 🙂 )

File: nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java [hunk 3]
```diff
@@ -362,11 +381,39 @@ public static void checkTypeParameterNullnessForFunctionReturnType(
...
+      Type.ArrayType lhsArrayType = (Type.ArrayType) lhsType;
+      Type.ArrayType rhsArrayType = (Type.ArrayType) rhsType;
+      Type lhsComponentType = lhsArrayType.getComponentType();
+      Type rhsComponentType = rhsArrayType.getComponentType();
+      boolean isLHSNullableAnnotated = isNullableAnnotated(lhsComponentType, state);
+      boolean isRHSNullableAnnotated = isNullableAnnotated(rhsComponentType, state);
+      // an array of @Nullable references is _not_ a subtype of an array of @NonNull references
+      if (isRHSNullableAnnotated && !isLHSNullableAnnotated) {
+        return false;
+      }
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java:412]: :+1:

**Review by lazaroclapp — APPROVED** [review]
LGTM!

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java (hunk 1 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java:232]: We don't do any inference for array contents nullability.  So when we declare `Integer[] x1 = ...` that means we are declaring that the array element type of `x1` is `@NonNull`.

File: nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java (hunk 2 above)
Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/jspecify/ArrayTests.java:351]: 075fa41ae62078cb740610a247184688e3030d18

**Pull request merged**